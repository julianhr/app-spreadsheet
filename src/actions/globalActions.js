import { createAction } from 'redux-starter-kit'


export const setActiveCell = createAction('SET_ACTIVE_CELL')
export const displayInputter = createAction('DISPLAY_INPUTTER')
