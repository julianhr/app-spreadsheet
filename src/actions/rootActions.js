import { createAction } from 'redux-starter-kit'


export const setActiveCell = createAction('SET_ACTIVE_CELL')

export function moveActiveCell(key) {
  return (dispatch, getState) => {
    const { activeCell, rows, columns } = getState()
    const currRow = activeCell.codePointAt(0) - 'a'.codePointAt(0)
    const currCol = Number.parseInt(activeCell[1]) - 1
    let nextRow = currRow
    let nextCol = currCol

    if (key === 'ArrowUp') { nextRow = currRow - 1 }
    else if (key === 'ArrowRight') { nextCol = currCol + 1 }
    else if (key === 'ArrowDown') { nextRow = currRow + 1 }
    else if (key === 'ArrowLeft') { nextCol = currCol - 1 }

    if (Math.min(nextRow, nextCol) >= 0 && nextRow < rows && nextCol < columns) {
      nextRow = String.fromCodePoint('a'.codePointAt(0) + nextRow)
      nextCol = nextCol + 1
      dispatch(setActiveCell(`${nextRow}${nextCol}`))
    }
  }
}
