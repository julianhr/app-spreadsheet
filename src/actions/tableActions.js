export function setCellData(location, entered, isEnteredValid, result) {
  return {
    type: 'SET_CELL_VALUE',
    payload: {
      location,
      data: {
        entered,
        result,
        isEnteredValid,
      }
    }
  }
}

export function clearCellData(location) {
  return (dispatch, getState) => {
    const value = getState().table[location]

    if (value === undefined) { return Promise.resolve() }

    let action = {
      type: 'DELETE_CELL_DATA',
      payload: location,
    }

    return Promise.resolve(dispatch(action))
  }
}
