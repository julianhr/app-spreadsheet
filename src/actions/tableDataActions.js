export function setCellData(location, entered, result) {
  return {
    type: 'SET_CELL_DATA',
    payload: { location, data: { entered, result } },
  }
}

export function clearCellData(location) {
  return (dispatch, getState) => {
    const value = getState().tableData[location]

    if (value === undefined) {
      return Promise.resolve()
    }

    return Promise.resolve(dispatch({
      type: 'DELETE_CELL_DATA',
      payload: location,
    }))
  }
}
