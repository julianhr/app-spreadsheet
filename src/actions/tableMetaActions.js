import { clamp } from '~/library/utils'
import {
  MIN_COL_WIDTH,
  MAX_COL_WIDTH,
  MIN_ROW_HEIGHT,
  MAX_ROW_HEIGHT,
} from '~/library/constants'


export function setColWidth(colLabel, width) {
  if (typeof width !== 'number') {
    throw new Error('width is not a number')
  }

  return {
    type: 'SET_COL_WIDTH',
    payload: { colLabel, width: clamp(width, MIN_COL_WIDTH, MAX_COL_WIDTH) }
  }
}

export function setRowHeight(rowLabel, height) {
  if (typeof height !== 'number') {
    throw new Error('height is not a number')
  }

  return {
    type: 'SET_ROW_HEIGHT',
    payload: { rowLabel, height: clamp(height, MIN_ROW_HEIGHT, MAX_ROW_HEIGHT) }
  }
}
