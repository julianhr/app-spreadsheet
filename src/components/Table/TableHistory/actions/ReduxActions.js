import reduxConnect from '../reduxConnect'


class ReduxActions {
  static reduxConnect = reduxConnect

  get reduxConnect() {
    return ReduxActions.reduxConnect
  }

  getCellData(location) {
    return this.reduxConnect.getState().tableData[location]
  }
}

export default ReduxActions
