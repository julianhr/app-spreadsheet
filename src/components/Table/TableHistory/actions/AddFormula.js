import ReduxActions from './ReduxActions'


class AddFormula extends ReduxActions {
  constructor(location, oldVertex, newVertex) {
    super()
    this.location = location
    this.entered = newVertex.entered
    this.result = newVertex.result
    this.previous = oldVertex && {
      entered: oldVertex.entered,
      result: oldVertex.result,
    }
  }

  execute() {
    this.reduxConnect.fn.setCellData(this.location, this.entered, this.result)
    this.reduxConnect.fn.setActiveCell(this.location)
  }

  unexecute() {
    if (this.previous) {
      const { entered, result } = this.previous
      this.reduxConnect.fn.setCellData(this.location, entered, result)
    } else {
      this.reduxConnect.fn.clearCellData(this.location)
    }

    this.reduxConnect.fn.setActiveCell(this.location)
  }
}

export default AddFormula
