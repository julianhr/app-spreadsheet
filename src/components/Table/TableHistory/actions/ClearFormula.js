import ReduxActions from './ReduxActions'


class ClearFormula extends ReduxActions {
  constructor(location) {
    super()
    this.previous = this.getCellData(location)
    this.location = location
  }

  execute() {
    this.reduxConnect.fn.clearCellData(this.location)
    this.reduxConnect.fn.setActiveCell(this.location)
  }

  unexecute() {
    if (!this.previous) { return }
    const { entered, result } = this.previous
    this.reduxConnect.fn.setCellData(this.location, entered, result)
    this.reduxConnect.fn.setActiveCell(this.location)
  }
}

export default ClearFormula