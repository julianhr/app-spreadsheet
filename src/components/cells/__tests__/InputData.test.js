import React from 'react'
import { create } from 'react-test-renderer'
import { render, fireEvent, cleanup, waitForElement } from 'react-testing-library'

import ConnectedInputData, {
  InputData,
  ERR_DIV_BY_ZERO,
  ERR_GENERIC,
} from '../InputData'
import MockApp from '~/__tests__/__mocks__/MockApp'
import appStoreGen from '~/reducers'
import { setCellValue } from '~/actions/tableActions'


const requiredProps = {
  // props
  replaceValue: false,
  location: 'B-2',
  onEscape: jest.fn(),
  onCommit: jest.fn(),
  // redux
  clearCellValue: jest.fn(),
  setCellValue: jest.fn(),
}

const testProps = {
  ...requiredProps,
  // props
  formula: 'test formula',
}

const createApp = (props) => {
  jest.spyOn(InputData.prototype, 'componentDidMount').mockReturnValueOnce('mocked cdm')
  const wrapper = create(<MockApp><InputData {...props} /></MockApp>)
  const { instance } = wrapper.root.findByType(InputData)
  return [wrapper, instance]
}

const renderApp = (props, customStore) => {
  const store = customStore || appStoreGen()
  const wrapper = render(
    <MockApp customStore={store}>
      <ConnectedInputData {...props} />
    </MockApp>
  )
  return [wrapper, store]
}

describe('InputData', () => {
  afterEach(cleanup)

  describe('props', () => {
    it('does not raise warning if al required props are passed', () => {
      expect(() => createApp(requiredProps)).not.toThrow()
    })

    it('raises warning if required props is not present', () => {
      Object.keys(requiredProps).forEach(key => {
        const props = {...testProps}
        delete props[key]
        expect(() => createApp(props)).toThrow()
      })
    })
  })

  describe('#focusInputTag', () => {
    it('focuses input tag when component mounts', () => {
      const [wrapper, _] = renderApp(testProps)
      const input = wrapper.getByDisplayValue('')
      expect(document.activeElement).toBe(input)
    })

    it('places cursor on far-right of formula string', async () => {
      const mockRef = { focus: jest.fn(), scrollLeft: 0, scrollWidth: 25 }
      const [_, instance] = createApp(testProps)
      instance.refInput.current = mockRef
  
      instance.focusInputTag()
      expect(instance.refInput.current.focus).toHaveBeenCalledTimes(1)
      expect(instance.refInput.current.scrollLeft).toEqual(mockRef.scrollWidth)
    })
  })

  describe('#setNewValue', () => {
    it('if string is empty the current value is deleted', async () => {
      const appStore = appStoreGen()
      const saved = { location: testProps.location, value: 'test string', formula: 'formula string' }
      await appStore.dispatch(setCellValue(saved))
      expect(appStore.getState().table[saved.location]).toBeTruthy()

      const [wrapper, _] = renderApp(testProps, appStore)
      const input = wrapper.getByDisplayValue(saved.formula)

      input.value = ''
      fireEvent.keyDown(input, { key: 'Enter' })
      expect(appStore.getState().table[saved.location]).toBeUndefined()
    })

    it('saves trimmed value if string is valid', () => {
      const appStore = appStoreGen()
      const saved = { location: testProps.location, value: 'test string', formula: 'formula string' }
      expect(appStore.getState().table[saved.location]).toBeUndefined()

      const [wrapper, _] = renderApp(testProps, appStore)
      const input = wrapper.getByDisplayValue('')
      const value = ' test value\n'

      input.value = value
      fireEvent.keyDown(input, { key: 'Enter' })
      expect(appStore.getState().table[saved.location].formula).toBe(value.trim())
    })

    it('calls #evaluateFormula if value starts with "="', () => {
      const [wrapper, _] = renderApp(testProps)
      const input = wrapper.getByDisplayValue('')
      const formula = '=2+3'

      jest.spyOn(InputData.prototype, 'evaluateFormula')
      input.value = formula
      fireEvent.keyDown(input, { key: 'Enter' })
      expect(InputData.prototype.evaluateFormula).toHaveBeenCalledTimes(1)
    })
  })

  describe('#evaluateFormula', () => {
    it('returns interpreted result if formula is valid', () => {
      const [_, store] = renderApp(testProps)
      const input = document.querySelector('#f-B-2')
      const formula = '=2+3'
      const cellData = { value: 5, formula }

      input.value = formula
      fireEvent.keyDown(input, { key: 'Enter' })
      expect(store.getState().table[testProps.location]).toEqual(cellData)
    })

    it('returns error if dividing by zero', () => {
      const [_, store] = renderApp(testProps)
      const input = document.querySelector('#f-B-2')
      const formula = '=3/0'
      const cellData = { value: ERR_DIV_BY_ZERO, formula }

      input.value = formula
      fireEvent.keyDown(input, { key: 'Enter' })
      expect(store.getState().table[testProps.location]).toEqual(cellData)
    })

    it('returns error if formula cannot be interpreted', () => {
      const [_, store] = renderApp(testProps)
      const input = document.querySelector('#f-B-2')
      const formula = '=3+5)'
      const cellData = { value: ERR_GENERIC, formula }

      input.value = formula
      fireEvent.keyDown(input, { key: 'Enter' })
      expect(store.getState().table[testProps.location]).toEqual(cellData)
    })
  })

  describe('#handleOnKeyDown', () => {
    test('sequence of events if key is "Escape"', () => {
      testProps.onEscape.mockReset()
      const [wrapper, _] = renderApp(testProps)
      const input = wrapper.getByDisplayValue('')
      fireEvent.keyDown(input, { key: 'Escape' })
      expect(testProps.onEscape).toHaveBeenCalledTimes(1)
    })

    test('sequence of event if key is "Enter', () => {
      testProps.onCommit.mockReset()
      const [wrapper, _] = renderApp(testProps)
      const input = wrapper.getByDisplayValue('')
      fireEvent.keyDown(input, { key: 'Enter' })
      expect(testProps.onCommit).toHaveBeenCalledTimes(1)
    })

    test('keydown event does not bubble unless key is "Enter" or "Escape"', () => {
      const events = 'b Enter 2 , t Escape E s 5'.split(' ')
      const spyKeyDown = jest.fn()
      const Listener = ({ children }) => <div onKeyDown={spyKeyDown}>{children}</div>
      render(
        <MockApp>
          <Listener>
            <InputData {...testProps}  />
          </Listener>
        </MockApp>
      )
      const input = document.querySelector('input')

      events.forEach(key => fireEvent.keyDown(input, { key }))
      expect(spyKeyDown).toHaveBeenCalledTimes(2)
    })
  })

  describe('#handleOnBlur', () => {
    test('stores value and calls #onCommit', () => {
      const [_, store] = renderApp(testProps)
      const formula = 'test value'
      const cellData = { value: formula, formula }

      testProps.onCommit.mockReset()
      const input = document.querySelector('input')
      input.value = formula
      fireEvent.blur(input)

      expect(store.getState().table[testProps.location]).toEqual(cellData)
      expect(testProps.onCommit).toHaveBeenCalledTimes(1)
    })
  })

  describe('functional tests', () => {
    const funcProps = {
      // props
      replaceValue: true,
      location: 'B-2',
      onEscape: jest.fn(),
      onCommit: jest.fn(),
    }

    it('displays default value if replaceValue is false', async () => {
      const cellData = {
        location: funcProps.location,
        value: 'test text',
        formula: 'test formula'
      }
      const store = appStoreGen()
      await store.dispatch(setCellValue(cellData))
      const props = {...funcProps}
      props.replaceValue = false
      const [wrapper, _] = renderApp(props, store)

      expect(wrapper.asFragment()).toMatchSnapshot()
    })

    it('will not display default value if replaceValue is true', async () => {
      const cellData = {
        location: funcProps.location,
        value: 'test text',
        formula: 'test formula'
      }
      const store = appStoreGen()
      await store.dispatch(setCellValue(cellData))
      const props = {...funcProps}
      props.replaceValue = true
      const [wrapper, _] = renderApp(props, store)

      expect(wrapper.asFragment()).toMatchSnapshot()
    })
  })
})
