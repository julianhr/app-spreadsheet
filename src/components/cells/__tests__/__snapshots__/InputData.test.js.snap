// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputData functional tests displays default value if replaceValue is false 1`] = `
<input
  class="css-p19rxp-Input ep2oxkq0"
  data-cell="input"
  data-location="B-2"
  value="test entered"
/>
`;

exports[`InputData functional tests will not display default value if replaceValue is true 1`] = `
<input
  class="css-p19rxp-Input ep2oxkq0"
  data-cell="input"
  data-location="B-2"
  value=""
/>
`;
